🚧 WIP 🚧  - Wrap CoreBluetooth to allow richer previews

MrSkwiggs opened the pull request:
This is my initial attempt at using Model that can be instantiated on the fly. 

While this sort-of works, most Views rely on `CoreBluetooth`-specific data to update their UI, which still makes it difficult to work with actual previews.

I think for this to work properly, I'd need to write protocols encompassing most of `CoreBluetooth`'s types (`CBManagerState`, `CBService`, `CBCharacteristic`, ...) so that I can then instantiate mocks (this should also help with writing tests later on). 

So something like:
- `protocol ManagerState`
- `protocol Service`
- `protocol Characteristic`
- `protocol Peripheral`
- `protocol Manager`

Then:
- Extends related `CoreBluetooth` types to make them conform to those protocols
- Write Mocks which also conform 
- Refactor views to rely on types which implement the protocols
- Refactor previews to use mocks
- ???
- profit

Let me know if this makes sense, it's quite a substantial effort but would definitely come in handy imo. If it does, I'll use this PR to work on this :)

artemnovichkov commented:
Sorry for the long answer, it's WWDC week :) I'll check it and return with the feedback.

MrSkwiggs commented:
Yeah haha, saw your tweets ;) no worries been busy on my end too anyways

artemnovichkov commented on ColorStripe/Extensions/Array+Convenience.swift:
```diff
@@ -0,0 +1,37 @@
+
+extension RandomAccessCollection {
+    func insertionIndex(for predicate: (Element) -> Bool) -> Index {
+        var slice : SubSequence = self[...]
```
It looks strange and... cool! 🤯

artemnovichkov commented on ColorStripe/Views/Device/DeviceView.swift:
```diff
@@ -55,9 +55,11 @@ struct DeviceView: View {
-                HStack {
-                    Text("Speed")
-                    Slider(value: $viewModel.state.speed, in: 0...1)
+                if let _ = viewModel.state.mode {
```
I prefer this syntax, because we don't use mode inside if statement
```suggestion
                if viewModel.state.mode != nil {
```

MrSkwiggs replied:
Done :)

artemnovichkov commented on ColorStripe/Views/Device/DeviceView.swift:
```diff
@@ -13,7 +13,7 @@ struct DeviceView: View {
     //MARK: - Lifecycle
     
-    init(peripheral: CBPeripheral) {
+    init(peripheral: PeripheralModel) {
```
We can remove extra `import CoreBluetooth` now in view models and views. Please check all cases.

MrSkwiggs replied:
Done

artemnovichkov commented on ColorStripe/Views/Device/DeviceViewModel.swift:
```diff
@@ -21,48 +21,58 @@ final class DeviceViewModel: ObservableObject {
     private lazy var cancellables: Set<AnyCancellable> = .init()
 
-    private let peripheral: CBPeripheral
+    private let peripheralModel: PeripheralModel
+    private var peripheral: CBPeripheral?
```
Why do you use both models?

MrSkwiggs replied:
Fixed with refactor :)

artemnovichkov commented on ColorStripe/Extensions/Array+Convenience.swift:
```diff
@@ -0,0 +1,37 @@
+
+import Foundation
+
+extension RandomAccessCollection {
```
[https://github.com/apple/swift-collections](swift-collections) has OrderedSet type, maybe it time to use third-party dependencies? What do you think? 🤔

MrSkwiggs replied:
with regards to my latest comment (debrief), because of the generic structure of the code, this will trickle down generics even further (ordered set requires the peripheral to be Hashable). Either that or I make the view model itself only accept hashable peripherals... It's all a big spaghetti of dependencies & generics 😓 and I'm not sure it should be this way... :/ If you have any suggestions or insight, would really appreciate it :D

artemnovichkov replied:
I think we can leave it as it is

artemnovichkov commented on ColorStripe/Views/Devices/PeripheralModel.swift:
```diff
@@ -0,0 +1,23 @@
+//
+//  PeripheralModel.swift
+//  ColorStripe
```
I prefer to remove class names and project name from headers, it's hard to support it in actual state. I'll add header template after PR merging.

MrSkwiggs replied:
Done

MrSkwiggs commented:
@artemnovichkov Thanks for the review. Before I jump into addressing your comments, could you please review my suggestion for the higher-level refactor ? Because I feel this will actually invalidate this PR altogether

artemnovichkov commented:
I've checked the whole PR and I agree with the refactoring. Initially, I didn't want to add extra abstractions, but it will help us to check the layout faster with previews and write tests as well.

MrSkwiggs commented:
Sorry for the delay, hectic at work! I'll give this a whirl when I have some time :)

MrSkwiggs commented:
Sorry for the force pushes, messed up my git config 😅 .
At any rate, this is so far what I managed to get working, the mocks still need a bit of work but at least, functionality-wise, nothing broke 🤓

artemnovichkov commented:
All good, great progress! Ping we when the PR will be ready for review

MrSkwiggs commented:
Alright, so with all of this done, here's a quick debrief of what went well and what could still do with improvements:

The **good**:
- The wrapping layer is correctly abstracting CoreBluetooth logic; it still works with CB while also allowing to write mock implementations;
- We can now instantiate mock data for our previews quite easily

The **bad**:
- Dat boilerplate doe
- Views now need to be generic to accommodate which data source is in use (actual BT or mock). Could be alleviated by using an intermediate struct that wraps only the values the view needs (title, id, etc.)

So all in all, it's a small win but could still do with some more work. 
I also need to still write some more mocks and overhaul a few more previews.

artemnovichkov commented:
Hi, @MrSkwiggs! Thanks for your work! I'm starting to the final changes 🥳

artemnovichkov commented on ColorStripe/CoreBluetooth Wrapping/Extensions/CBCentralManager+Wrapping.swift:
```diff
@@ -0,0 +1,51 @@
+import Foundation
+import CoreBluetooth
+
+class CentralManagerWrapper: NSObject, ManagerWrapper, CBCentralManagerDelegate {
```
```suggestion
final class CentralManagerWrapper: NSObject, ManagerWrapper, CBCentralManagerDelegate {
```

artemnovichkov commented on ColorStripe/CoreBluetooth Wrapping/Extensions/CBCentralManager+Wrapping.swift:
```diff
@@ -0,0 +1,51 @@
+    weak var delegateWrapper: ManagerDelegateWrapper?
+    
+    var stateWrapper: ManagerState {
+            .init(from: cbManager.state)
```
```suggestion
        .init(from: cbManager.state)
```

MrSkwiggs replied:
heh, automatic indentation from XCode 13 Beta 1 :d

artemnovichkov commented on ColorStripe/CoreBluetooth Wrapping/Extensions/CBCentralManager+Wrapping.swift:
```diff
@@ -0,0 +1,51 @@
+            .init(delegate: delegate, queue: queue)
+    }
+    
+    func scanForPeripheralsWrapper(withServices serviceUUIDs: [CBUUID]?) {
```
Why did you add `Wrapper` word to some functions?

MrSkwiggs replied:
Somehow when I kept the native func name, I ran into issues when invoking them and wanting to specify either the native implementation or the wrapper implementation. Using that suffix helped there. But might also be related to XCode 13 Beta

artemnovichkov commented on ColorStripe/CoreBluetooth Wrapping/Extensions/CBCentralManager+Wrapping.swift:
```diff
@@ -0,0 +1,51 @@
+    }
+    
+    static func initialize(delegate: ManagerDelegateWrapper?, queue: DispatchQueue) -> Self {
+            .init(delegate: delegate, queue: queue)
```
```suggestion
        .init(delegate: delegate, queue: queue)
```

artemnovichkov commented on ColorStripe/CoreBluetooth Wrapping/Extensions/CBPeripheral+Wrapping.swift:
```diff
@@ -0,0 +1,34 @@
+    }
+    
+    func discoverCharacteristicsWrapper(_ characteristicUUIDs: [CBUUID]?, for service: ServiceWrapper) {
+        discoverCharacteristics(characteristicUUIDs, for: service as! CBService)
```
Is it possible to avoid force upwraps?

artemnovichkov commented on ColorStripe/CoreBluetooth Wrapping/ManagerState.swift:
```diff
@@ -0,0 +1,42 @@
+import Foundation
+import CoreBluetooth
+
+enum ManagerState {
+    case poweredOn
+    case poweredOff
+    case resetting
+    case unauthorized
+    case unknown
+    case unsupported
+}
```
In my opinion, we can use CBManagerState to prevent code duplication

MrSkwiggs replied:
Fair enough actually, this indeed brings nothing and I'm already importing `CoreBluetooth` anyway because of CBUUIDs

artemnovichkov commented on ColorStripe/CoreBluetooth Wrapping/Mocks/MockManager.swift:
```diff
@@ -0,0 +1,81 @@
+import Foundation
+import CoreBluetooth
+
+class MockManager: ManagerWrapper {
```
Please add final keywords where it's possible
```suggestion
final class MockManager: ManagerWrapper {
```

artemnovichkov commented on ColorStripe/Views/Devices/DevicesViewModel.swift:
```diff
@@ -30,12 +38,19 @@ final class DevicesViewModel: ObservableObject {
+        
         manager.peripheralSubject
-            .filter { [weak self] in self?.peripherals.contains($0) == false }
+//            .filter { [weak self] in self?.peripherals.contains($0) == false }
+//            .sink { [weak self] in
+//                self?.peripherals.insertSorted($0)
+//            }
```
Why did you leave these comments?

MrSkwiggs replied:
Well I'd still like to implement some level of sorting here... But yeah I agree the code comment should be removed, will do so when I get some time to work some more on this

artemnovichkov commented on ColorStripe/CoreBluetooth Wrapping/ManagerWrapper.swift:
```diff
@@ -0,0 +1,21 @@
+
+protocol ManagerWrapper {
+    
+    static func initialize(delegate: ManagerDelegateWrapper?, queue: DispatchQueue) -> Self
```
can we replace this initialization with this one?
```swift
init(delegate: ManagerDelegateWrapper?, queue: DispatchQueue)
```

MrSkwiggs replied:
No, it's not possible to add a convenience init for `CBCentralManager` unfortunately...

artemnovichkov commented on ColorStripe/Views/Devices/DevicesView.swift:
```diff
@@ -5,25 +5,23 @@
 import CoreBluetooth
 
-struct DevicesView: View {
+struct DevicesView<T: ManagerWrapper>: View {
```
I prefer to use more descriptive names for generics
```suggestion
struct DevicesView<Manager: ManagerWrapper>: View {
```

MrSkwiggs commented:
Thanks for the review @artemnovichkov ! Will get back to it as soon as I've got some time :)
